### PR TITLE
♻️ Use std::function instead of mbed::Callback

### PR DESCRIPTION
**Pull request by YannLocatelli**
# Validations

- [x] BLE
- [x] Battery - onChargeDidStart
- [x] Battery - onChargeDidStop
- [x] RFID

### CODE REVIEW

File: drivers/CoreBattery/source/CoreBattery.cpp [hunk 1]
```diff
@@ -4,18 +4,22 @@
...
 #include "MathUtils.h"
 
 namespace leka {
 
-void CoreBattery::onChargeDidStart(mbed::Callback<void()> const &callback)
+void CoreBattery::onChargeDidStart(std::function<void()> const &callback)
 {
-	_charge_status_input.rise(callback);
+	_on_charge_did_start = callback;
+	_charge_status_input.rise(mbed::Callback<void()> {[this] { _on_charge_did_start(); }});
```
Comment (ladislas) [inline drivers/CoreBattery/source/CoreBattery.cpp:16]: is it not possible to do just that? Do we really need to go from std::function to lambda to mbed::callback?

```suggestion
	_charge_status_input.rise([this] { _on_charge_did_start(); });
```

File: drivers/CoreBattery/source/CoreBattery.cpp [hunk 2]
```diff
@@ -4,18 +4,22 @@
...
 }
 
-void CoreBattery::onChargeDidStop(mbed::Callback<void()> const &callback)
+void CoreBattery::onChargeDidStop(std::function<void()> const &callback)
 {
-	_charge_status_input.fall(callback);
+	_on_charge_did_stop = callback;
+	_charge_status_input.fall(mbed::Callback<void()> {[this] { _on_charge_did_stop(); }});
```
Comment (ladislas) [inline drivers/CoreBattery/source/CoreBattery.cpp:22]: same as above if possible

```suggestion
	_charge_status_input.fall([this] { _on_charge_did_stop(); });
```

File: drivers/CoreBufferedSerial/source/CoreBufferedSerial.cpp [hunk 3]
```diff
@@ -33,9 +33,10 @@ void CoreBufferedSerial::disable_input()
 	_serial.enable_input(false);
 }
 
-void CoreBufferedSerial::sigio(mbed::Callback<void()> func)
+void CoreBufferedSerial::sigio(std::function<void()> const &callback)
 {
-	_serial.sigio(func);
+	_sigio_callback = callback;
+	_serial.sigio(mbed::Callback<void()> {[this] { _sigio_callback(); }});
```
Comment (ladislas) [inline drivers/CoreBufferedSerial/source/CoreBufferedSerial.cpp:39]: as we do in `CoreRFIDReaderCR95HF` mbed::callback could be removed

```suggestion
	_serial.sigio([this] { _sigio_callback(); });
```

File: libs/BLEKit/source/BLEKit.cpp [hunk 4]
```diff
@@ -31,7 +33,7 @@ void BLEKit::init()
 
 void BLEKit::processEvents(BLE::OnEventsToProcessCallbackContext *context)
 {
-	_event_queue.callMbedCallback(mbed::callback(&context->ble, &BLE::processEvents));
+	_event_queue.call(mbed::Callback {&context->ble, &BLE::processEvents});
```
Comment (ladislas) [inline libs/BLEKit/source/BLEKit.cpp:36]: also here, is mbed::Callback really necessary compared to a lambda?

**Review by ladislas — CHANGES_REQUESTED** [review]
thanks for this PR! good to see mbed::callback going away in a lot of places :)

I've added a few suggestions where I think that mbed::callback can be completely removed and replaced by simple lambdas.

**Review by HPezz — COMMENTED** [review]
Nice 👍 

- [x] Validated on robot

**Review by ladislas — APPROVED** [review]

**Pull request merged**